File: api/src/models/ServiceLog.ts
import mongoose, { Schema, Types, Document } from 'mongoose';

// Interface for ServiceLog document
export interface IServiceLog extends Document {
  vehicleId: Types.ObjectId;
  date: Date;
  mileage: number;
  serviceType: string;
  description?: string;
  mechanic?: string;
  totalCost: number;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

// Interface for static methods
interface IServiceLogModel extends mongoose.Model<IServiceLog> {
  findByVehicle(
    vehicleId: string,
    options?: { page?: number; limit?: number }
  ): Promise<IServiceLog[]>;
  findByVehicleAndId(vehicleId: string, serviceLogId: string): Promise<IServiceLog | null>;
  countByVehicle(vehicleId: string): Promise<number>;
}

// Schema definition
const ServiceLogSchema = new Schema<IServiceLog, IServiceLogModel>(
  {
    vehicleId: {
      type: Schema.Types.ObjectId,
      ref: 'Vehicle',
      required: [true, 'Vehicle ID is required'],
      index: true,
    },
    date: {
      type: Date,
      required: [true, 'Date is required'],
      default: Date.now,
      index: true,
    },
    mileage: {
      type: Number,
      required: [true, 'Mileage is required'],
      min: [0, 'Mileage cannot be negative'],
    },
    serviceType: {
      type: String,
      required: [true, 'Service type is required'],
      trim: true,
      maxlength: [100, 'Service type cannot exceed 100 characters'],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [500, 'Description cannot exceed 500 characters'],
      default: undefined,
    },
    mechanic: {
      type: String,
      trim: true,
      maxlength: [100, 'Mechanic name cannot exceed 100 characters'],
      default: undefined,
    },
    totalCost: {
      type: Number,
      required: [true, 'Total cost is required'],
      min: [0, 'Total cost cannot be negative'],
    },
    notes: {
      type: String,
      trim: true,
      maxlength: [500, 'Notes cannot exceed 500 characters'],
      default: undefined,
    },
  },
  {
    timestamps: true,
  }
);

// Compound index for vehicleId + date (for efficient querying and sorting)
ServiceLogSchema.index({ vehicleId: 1, date: -1 });

// Statics (model-level)
ServiceLogSchema.statics.findByVehicle = function (
  vehicleId: string,
  options?: { page?: number; limit?: number }
) {
  let query = this.find({ vehicleId }).sort({ date: -1 });

  if (options && options.page !== undefined && options.limit !== undefined) {
    const skip = (options.page - 1) * options.limit;
    query = query.skip(skip).limit(options.limit);
  }

  return query;
};

ServiceLogSchema.statics.findByVehicleAndId = function (
  vehicleId: string,
  serviceLogId: string
) {
  return this.findOne({ _id: serviceLogId, vehicleId });
};

ServiceLogSchema.statics.countByVehicle = function (vehicleId: string) {
  return this.countDocuments({ vehicleId });
};

// Virtuals
ServiceLogSchema.virtual('id').get(function () {
  return this._id.toHexString();
});

// Transform JSON output
ServiceLogSchema.set('toJSON', {
  virtuals: true,
  transform: function (_doc, ret) {
    const transformed = { ...ret } as Record<string, unknown>;
    delete transformed._id;
    delete transformed.__v;
    return transformed;
  },
});

export default mongoose.model<IServiceLog, IServiceLogModel>('ServiceLog', ServiceLogSchema);

File: mobile/lib/domain/entities/service_log.dart
class ServiceLog {
  final String id;
  final String vehicleId;
  final DateTime date;
  final int mileage;
  final String serviceType;
  final String? description;
  final String? mechanic;
  final double totalCost;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ServiceLog({
    required this.id,
    required this.vehicleId,
    required this.date,
    required this.mileage,
    required this.serviceType,
    this.description,
    this.mechanic,
    required this.totalCost,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ServiceLog.fromJson(Map<String, dynamic> json) => ServiceLog(
        id: json['id'] as String,
        vehicleId: json['vehicleId'] as String,
        date: DateTime.parse(json['date'] as String),
        mileage: json['mileage'] as int,
        serviceType: json['serviceType'] as String,
        description: json['description'] as String?,
        mechanic: json['mechanic'] as String?,
        totalCost: (json['totalCost'] as num).toDouble(),
        notes: json['notes'] as String?,
        createdAt: DateTime.parse(json['createdAt'] as String),
        updatedAt: DateTime.parse(json['updatedAt'] as String),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'vehicleId': vehicleId,
        'date': date.toIso8601String(),
        'mileage': mileage,
        'serviceType': serviceType,
        'description': description,
        'mechanic': mechanic,
        'totalCost': totalCost,
        'notes': notes,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  ServiceLog copyWith({
    String? id,
    String? vehicleId,
    DateTime? date,
    int? mileage,
    String? serviceType,
    String? description,
    String? mechanic,
    double? totalCost,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) =>
      ServiceLog(
        id: id ?? this.id,
        vehicleId: vehicleId ?? this.vehicleId,
        date: date ?? this.date,
        mileage: mileage ?? this.mileage,
        serviceType: serviceType ?? this.serviceType,
        description: description ?? this.description,
        mechanic: mechanic ?? this.mechanic,
        totalCost: totalCost ?? this.totalCost,
        notes: notes ?? this.notes,
        createdAt: createdAt ?? this.createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
      );
}